opencensus/trace/span.py

```diff
@@ -12,20 +12,113 @@
+    from collections.abc import Sequence
+except ImportError:
+    from collections import MutableMapping
+    from collections import Sequence
+
+from collections import OrderedDict
+from collections import deque
 from datetime import datetime
 from itertools import chain
 
 from opencensus.common import utils
-from opencensus.trace import attributes
+from opencensus.trace import attributes as attributes_module
 from opencensus.trace import base_span
 from opencensus.trace import link as link_module
-from opencensus.trace import stack_trace
-from opencensus.trace import status
-from opencensus.trace import time_event as time_event_module
+from opencensus.trace import stack_trace as stack_trace_module
+from opencensus.trace import status as status_module
+from opencensus.trace import time_event
 from opencensus.trace.span_context import generate_span_id
 from opencensus.trace.tracers import base
 
 
+# https://github.com/census-instrumentation/opencensus-specs/blob/master/trace/TraceConfig.md  # noqa
+MAX_NUM_ATTRIBUTES = 32
+MAX_NUM_ANNOTATIONS = 32
+MAX_NUM_MESSAGE_EVENTS = 128
+MAX_NUM_LINKS = 32
+
+
+class BoundedList(Sequence):
+    """An append only list with a fixed max size."""
+    def __init__(self, maxlen):
+        self.dropped = 0
+        self._dq = deque(maxlen=maxlen)
+
+    def __repr__(self):
+        return repr(self._dq)
+
+    def __getitem__(self, index):
+        return self._dq[index]
+
+    def __len__(self):
+        return len(self._dq)
+
+    def __iter__(self):
+        return iter(self._dq)
+
+    def append(self, item):
+        if len(self._dq) == self._dq.maxlen:
+            self.dropped += 1
+        self._dq.append(item)
+
+    def extend(self, seq):
+        to_drop = len(seq) + len(self._dq) - self._dq.maxlen
+        if to_drop > 0:
+            self.dropped += to_drop
+        self._dq.extend(seq)
+
+    @classmethod
+    def from_seq(cls, maxlen, seq):
+        if len(seq) > maxlen:
+            raise ValueError
+        bounded_list = cls(maxlen)
+        bounded_list._dq = deque(seq, maxlen=maxlen)
+        return bounded_list
+
+
+class BoundedDict(MutableMapping):
+    """A dict with a fixed max capacity."""
+    def __init__(self, maxlen):
+        self.maxlen = maxlen
+        self.dropped = 0
+        self._dict = OrderedDict()
+
+    def __repr__(self):
+        return repr(self._dict)
+
+    def __getitem__(self, key):
+        return self._dict[key]
+
+    def __setitem__(self, key, value):
+        if key in self._dict:
+            del self._dict[key]
+        elif len(self._dict) == self.maxlen:
+            del self._dict[next(iter(self._dict.keys()))]
+            self.dropped += 1
+        self._dict[key] = value
+
+    def __delitem__(self, key):
+        del self._dict[key]
+
+    def __iter__(self):
+        return iter(self._dict)
+
+    def __len__(self):
+        return len(self._dict)
+
+    @classmethod
+    def from_map(cls, maxlen, mapping):
+        if len(mapping) > maxlen:
```

Comment:
Do we want to provide thread-safety for classmethod? If yes, probably consider a shallow copy before determine the size:
```
mapping = OrderedDict(mapping)
if len(mapping) > maxlen:
    ...
```

Reply:
AFAICT that's not exactly threadsafe either since it's still possible to modify `mapping` while `MutableMapping.update` is iterating over it to make the copy, but that's certainly better than it being modified between the length check and the copy.